src/main/java/org/jitsi/videobridge/Endpoint.java

```diff
@@ -155,18 +149,8 @@
      * prematurely, or more than once.
      *
      */
-    private DtlsTransport transportManager;
-
-    /**
-     * The exception thrown from the attempt to initialize the transport manager,
-     * if any.
-     */
-    private IOException transportManagerException = null;
-
-    /**
-     * Synchronizes access to {@link #transportManager}.
-     */
-    private final Object transportManagerSyncRoot = new Object();
+    private final CompletableFuture<DtlsTransport> dtlsTransportFuture
```

Comment:
I'll try to explain my motivation behind the change and hope those 2 question will be implicitly answered.

Currently in [`master`](https://github.com/jitsi/jitsi-videobridge/commit/6afe06c99f2e4046bff409e6bb80631330b26a32) initialization of `DtlsTransport` is handled by 4 fields:
1. [`transportManager`](https://github.com/jitsi/jitsi-videobridge/blob/6afe06c99f2e4046bff409e6bb80631330b26a32/src/main/java/org/jitsi/videobridge/Endpoint.java#L158)
2. [`transportManagerException`](https://github.com/jitsi/jitsi-videobridge/blob/6afe06c99f2e4046bff409e6bb80631330b26a32/src/main/java/org/jitsi/videobridge/Endpoint.java#L164)
3. [`transportManagerSyncRoot`](https://github.com/jitsi/jitsi-videobridge/blob/6afe06c99f2e4046bff409e6bb80631330b26a32/src/main/java/org/jitsi/videobridge/Endpoint.java#L169)
4. [`onTransportManagerSet`](https://github.com/jitsi/jitsi-videobridge/blob/6afe06c99f2e4046bff409e6bb80631330b26a32/src/main/java/org/jitsi/videobridge/Endpoint.java#L129)

I've tried to understand motivation behind current implementation and deduced the following motivation:
1. `transportManager` field is necessary to hold reference to `DtlsTransport` instance once it successfully initialized;
2. `transportManagerException` field is necessary to remember the exception which happen during initialization of `DtlsTransport` is it happen;
3. `transportManagerSyncRoot` since there is basically two fields which handle the state of single logical thing - initialization of `DtlsTransport` - it necessary to synchronize access to both variables, such that only one of them is actually holding a value.
4. `onTransportManagerSet` - serves a kind of event that `DtlsTransport` is successfully initialized, basically it is a completion handler of successful `DtlsTransport` initialization.

With that understanding (which might be incorrect, because it was inferred from code, not from conversation with author of original code) I realized that this four fields is basically reinvent `CompletableFuture`.
`CompletableFuture` could either hold successfully computed value or exception happen during value computation, it may only be set once, it provides thread safety - so no external synchronization required and it may have completion handlers. All this functionality with single instance of `CompletableFuture`!
The only possible drawback of that is that whenever it necessary to read `DtlsTransport` it necessary to examine the state of `CompletableFuture`, but that's easily overcame with special getter, which in the proposed PR is [`tryGetDtlsTransport`](https://github.com/mstyura/jitsi-videobridge/blob/802f7f4b6605a1a472c29916f811fb28990e0ab6/src/main/java/org/jitsi/videobridge/Endpoint.java#L199).
On the other hand the explicit necessity to check `CompletableFuture` might be consider as a good thing, because it forces user of `DtlsTransport` to explicitly check the state of `DtlsTransport` initialization. For example, I think I've fixed [one potential NPE](https://github.com/mstyura/jitsi-videobridge/blob/802f7f4b6605a1a472c29916f811fb28990e0ab6/src/main/java/org/jitsi/videobridge/Endpoint.java#L739-L743) due to this.

I hope this explanation will give you more insight on my motivation behind this change and is answered the questions you've raised. 
Thanks a lot for review, Brian!

Reply:
thanks for the great explanation, @mstyura, you've helped refresh my memory a bit and i think your assessment of replacing the current scheme with `CompletableFuture` makes sense.  looking at the old code again, i'm wondering if we can clean some things up here to improve this.  one thing i'm thinking when looking at this is why we can't just set up the transport in the endpoint constructor (and throw if it fails).  this would:
1) remove the ambiguity about when it is set
2) remove the need to 'remember' the thrown exception, as we'd (rightfully) prevent the endpoint from being created
3) remove the error checking in endpoint about retrieving it and it potentially being null

now, i _think_ i've had this thought before and there was a reason we couldn't, but perhaps in 2.0 that reason is now gone.  do you know, based on what you've seen, if that's not possible for some reason? i can look into it as well but would take some time before i'd be able.

Reply:
>  do you know, based on what you've seen, if that's not possible for some reason? i can look into it as well but would take some time before i'd be able.

I'll investigate this option tomorrow and let you know the result.

Reply:
> i'm wondering if we can clean some things up here to improve this. one thing i'm thinking when looking at this is why we can't just set up the transport in the endpoint constructor (and throw if it fails).

That would be ideal and today I did necessary refactorings to achieve the above stated goal.
It's occurred to be doable in `JVB 2.0`

With latest changes to this PR the following major things has been done:
1. [`DtlsTransport` is now initialized](https://github.com/mstyura/jitsi-videobridge/blob/d8a90777486b03e99109fda1c0a1494dc80aceec/src/main/java/org/jitsi/videobridge/Endpoint.java#L253) inside `Endpoint` constructor with initial `ICE` role specified in advance;
2. [`Conference.getOrCreateLocalEndpoint`](https://github.com/jitsi/jitsi-videobridge/blob/6afe06c99f2e4046bff409e6bb80631330b26a32/src/main/java/org/jitsi/videobridge/Conference.java#L822) has been replaced with 2 dedicated methods [`getLocalEndpoint`](https://github.com/mstyura/jitsi-videobridge/blob/d8a90777486b03e99109fda1c0a1494dc80aceec/src/main/java/org/jitsi/videobridge/Conference.java#L826) and [`createLocalEndpoint`](https://github.com/mstyura/jitsi-videobridge/blob/d8a90777486b03e99109fda1c0a1494dc80aceec/src/main/java/org/jitsi/videobridge/Conference.java#L648).
3. During `ColibriConferenceIQ` handling all endpoints allocated and initialized [in the single place](https://github.com/mstyura/jitsi-videobridge/blob/d8a90777486b03e99109fda1c0a1494dc80aceec/src/main/java/org/jitsi/videobridge/shim/VideobridgeShim.java#L315);

@bbaldino could you please review latest changes? Thanks a lot in advance!